// react-scroller/dist/Scroller.js
var scrollerComponent;

scrollerComponent = function(React) {
  return React.createClass({
    currentPage: 1,
    getDefaultProps: function() {
      return {
        page: 1,
        "continue": false,
        load: function() {},
        threshold: 100,
        container: window
      };
    },
    componentDidMount: function() {
      this.currentPage = this.props.page;
      return this.props.load(this.currentPage);
    },
    componentDidUpdate: function() {
      return this.addListener();
    },
    componentWillUnmount: function() {
      return this.removeListener();
    },
    componentWillReceiveProps: function(nextProps) {
      if (nextProps.page) {
        return this.currentPage = nextProps.page;
      }
    },
    render: function() {
      return React.createElement('div', null, this.props.children);
    },
    getContainerOffset: function() {
      var height, scroll;
      scroll = this.props.container.pageYOffset || this.props.container.scrollTop;
      return height = this.props.container.innerHeight || this.props.container.offsetHeight;
    },
    getContainerHeight: function() {
      return this.props.container.innerHeight || this.props.container.offsetHeight;
    },
    getContainerScrollPos: function() {
      return this.props.container.pageYOffset || this.props.container.scrollTop;
    },
    scrollListener: function() {
      var containerOffset, scroller, scrollerOffset;
      scroller = React.findDOMNode(this);
      scrollerOffset = scroller.scrollTop + scroller.offsetHeight;
      containerOffset = this.getContainerScrollPos() + this.getContainerHeight();
      if ((scrollerOffset - containerOffset) < Number(this.props.threshold)) {
        this.removeListener();
        return this.props.load(++this.currentPage);
      }
    },
    addListener: function() {
      if (!(this.props["continue"] && this.props.container)) {
        return;
      }
      this.props.container.addEventListener('scroll', this.scrollListener);
      this.props.container.addEventListener('resize', this.scrollListener);
      return this.scrollListener();
    },
    removeListener: function() {
      if (!this.props.container) {
        return;
      }
      this.props.container.removeEventListener('scroll', this.scrollListener);
      return this.props.container.removeEventListener('resize', this.scrollListener);
    }
  });
};

module.exports = scrollerComponent;
